contracts/ExitHandler.sol

```diff
@@ -29,42 +29,392 @@ contract ExitHandler is DepositHandler {
+      bytes32 previousHash;
+      bytes memory previousTxData;
+      uint64 previousPos;
+      LimboChallenge memory prevChallenge = limboExit.challenge;
+      bytes memory _prevProof = prevChallenge.proof;
+      (previousPos, previousHash, previousTxData) = TxLib.validateProof(32 * (_youngestInputProof.length + 2) + 64, _prevProof);
+      TxLib.Tx memory previousTx = TxLib.parseTx(previousTxData);
+      uint32 prevHeight = bridge.periods[_prevProof[0]].height;
+      
+      require(blockHeight < prevHeight);
+      // pay challengeStake to winner from exitHandler
+      tokens[transferTx.outs[0].color].addr.transferFrom(address(this), msg.sender, challengeStake);
+
+    }
+    limboExit.isValid = false;
+    limboExit.input[InputNo].exitable = false;
+    limboExit.challenge = LimboChallenge({
+      owner: msg.sender,
+      resolved: false,
+      proof: _spendProof
+    });
+  }
+
+  function challengeLimboExitByOutputSpend(
+    bytes32 exitId,
+    bytes memory inTxData, uint8 OutputNo,
+    bytes memory _youngestInputProof, bytes memory _spendProof, uint8 InputNo) 
+    public payable {
+    require(msg.value >= challengeStake, "Not enough ether sent to challenge exit");
+    LimboExit memory limboExit = limboExits[exitId];
+    bytes32 inTxHash = keccak256(inTxData);
+
+    bytes32 competingHash;
+    bytes memory competingTxData;
+    uint64 competingPos;
+    
+    TxLib.Tx memory transferTx = TxLib.parseTx(inTxData);
+    (competingPos, competingHash, competingTxData) = TxLib.validateProof(32 * (_youngestInputProof.length + 2) + 64, _spendProof);
+    TxLib.Tx memory competingTx = TxLib.parseTx(competingTxData);
+
+    uint32 blockHeight = bridge.periods[_spendProof[0]].height;
+
+    if (limboExit.isValid == true){
+      uint256 offset = uint8(uint256(_spendProof[1] >> 248));
+      // getting spendTx input address
+      address owner = TxLib.recoverTxSigner(offset, _spendProof);      
+      LimboOut memory inOutput = limboExit.output[OutputNo];
+      TxLib.Input memory competingInput = competingTx.input[InputNo];
+
+      // check if same or not ?      
+      require(inOutput.owner == owner);
+      require(inOutput.amount == competingTx.outs[competingInput.outpoint.pos]);
+
+    } else {
+      // if someone challenged it before and became successful
+      LimboChallenge memory prevChallenge = limboExit.challenge;
+      bytes32[] memory _prevProof = prevChallenge.proof;
+      uint64 previousPos;
+      bytes32 previousHash;
+      bytes memory previousTxData;
+
+      (previousPos, previousHash, previousTxData) = TxLib.validateProof(32 * (_youngestInputProof.length + 2) + 64, _prevProof);
+      TxLib.Tx memory previousTx = TxLib.parseTx(previousTxData);      
+      uint32 prevHeight = bridge.periods[_prevProof[0]].height;
+
+      require(blockHeight < prevHeight);
+      tokens[transferTx.outs[0].color].addr.transferFrom(address(this), msg.sender, challengeStake);
```

Comment:
`challengeStake` is in ether, so it should transfer ether here. Also the destination is not correct: we are returning stake of the previous challenger, so it should transfer to `prevChallenge.owner`